File: predictive-epidemic-modeling-platform (1)/src/engine/seirSolver.ts
/**
 * SEIR Differential Equations Solver
 * 
 * Extended SEIR model with mortality and vaccination:
 *   dS/dt = -β*S*I/N - vaccine_rate*S
 *   dE/dt = β*S*I/N - σ*E
 *   dI/dt = σ*E - γ*I - μ*I
 *   dR/dt = γ*I + vaccine_rate*S*(1 - vaccine_efficacy)
 *   dD/dt = μ*I
 *
 * β is modulated by temperature, humidity, mask compliance, and isolation index.
 * Uses 4th-order Runge-Kutta (RK4) integration for numerical accuracy.
 */

import { SEIRState, SimulationParams } from '../types';

// RK4 integration step for SEIR equations
function rk4Step(
  state: SEIRState,
  dt: number,
  params: SimulationParams,
  N: number,
  currentBeta: number,
  mobilityIn: number,
  mobilityOut: number
): SEIRState {
  const { sigma, gamma, mu, vaccineRate, vaccineEfficacy } = params;
  const vRate = (vaccineRate / 100) * 0.01; // convert percent to daily rate

  const deriv = (s: SEIRState): SEIRState => {
    const totalPop = N;
    const effectiveBeta = currentBeta;
    const contactTerm = (effectiveBeta * s.S * s.I) / totalPop;
    const vaccinationTerm = vRate * s.S;
    const ve = vaccineEfficacy / 100;

    return {
      S: -contactTerm - vaccinationTerm - mobilityOut + mobilityIn,
      E: contactTerm - sigma * s.E,
      I: sigma * s.E - gamma * s.I - mu * s.I,
      R: gamma * s.I + vaccinationTerm * (1 - ve),
      D: mu * s.I,
    };
  };

  const addState = (a: SEIRState, b: SEIRState, factor: number): SEIRState => ({
    S: a.S + b.S * factor,
    E: a.E + b.E * factor,
    I: a.I + b.I * factor,
    R: a.R + b.R * factor,
    D: a.D + b.D * factor,
  });

  const k1 = deriv(state);
  const s2 = addState(state, k1, dt / 2);
  // Clamp negatives for deriv evaluation
  const s2Clamped = clampState(s2);
  const k2 = deriv(s2Clamped);
  const s3 = addState(state, k2, dt / 2);
  const s3Clamped = clampState(s3);
  const k3 = deriv(s3Clamped);
  const s4 = addState(state, k3, dt);
  const s4Clamped = clampState(s4);
  const k4 = deriv(s4Clamped);

  return {
    S: state.S + (dt / 6) * (k1.S + 2 * k2.S + 2 * k3.S + k4.S),
    E: state.E + (dt / 6) * (k1.E + 2 * k2.E + 2 * k3.E + k4.E),
    I: state.I + (dt / 6) * (k1.I + 2 * k2.I + 2 * k3.I + k4.I),
    R: state.R + (dt / 6) * (k1.R + 2 * k2.R + 2 * k3.R + k4.R),
    D: state.D + (dt / 6) * (k1.D + 2 * k2.D + 2 * k3.D + k4.D),
  };
}

function clampState(state: SEIRState): SEIRState {
  return {
    S: Math.max(0, state.S),
    E: Math.max(0, state.E),
    I: Math.max(0, state.I),
    R: Math.max(0, state.R),
    D: Math.max(0, state.D),
  };
}

// Compute temperature/humidity-modulated transmission rate
function computeBeta(
  baseBeta: number,
  avgTemp: number,
  humidity: number,
  maskCompliance: number,
  isolationIndex: number
): number {
  // Temperature effect: moderate temps increase transmission
  // Peak transmission around 10-15°C, decreases at extremes
  const tempFactor = 1 + 0.02 * Math.exp(-((avgTemp - 12) ** 2) / 100);

  // Humidity effect: moderate humidity (40-60%) increases transmission
  const humidityFactor = 1 + 0.1 * Math.exp(-((humidity - 50) ** 2) / 2000);

  // Mask compliance reduces effective contact
  const maskFactor = 1 - (maskCompliance / 100) * 0.6; // Max 60% reduction

  // Isolation reduces mixing
  const isolationFactor = 1 - isolationIndex * 0.7; // Max 70% reduction

  return Math.max(0.01, baseBeta * tempFactor * humidityFactor * maskFactor * isolationFactor);
}

/**
 * Solve single-node SEIR model
 */
export function solveSingleNodeSEIR(
  population: number,
  initialSEIR: SEIRState,
  params: SimulationParams,
  days: number,
  avgTemp: number,
  humidity: number,
  mobilityFlows: number[][] // [[day, inFlow], ...]
): SEIRState[] {
  const result: SEIRState[] = [];
  let state = { ...initialSEIR };

  for (let day = 0; day <= days; day++) {
    result.push({ ...state });

    if (day >= days) break;

    // Get mobility for this day
    const mob = mobilityFlows.find(([d]) => d === day) || [0, 0];
    const mobilityIn = mob[1];
    const mobilityOut = 0; // computed in network layer

    const currentBeta = computeBeta(
      params.beta,
      avgTemp,
      humidity,
      params.maskCompliance,
      params.isolationIndex
    );

    // Sub-stepping for accuracy (4 steps per day)
    const dt = 0.25;
    for (let step = 0; step < 4; step++) {
      state = rk4Step(state, dt, params, population, currentBeta, mobilityIn / 4, mobilityOut / 4);
    }

    state = clampState(state);

    // Ensure population doesn't grow beyond initial
    const total = state.S + state.E + state.I + state.R + state.D;
    if (total > population) {
      const scale = population / total;
      state = {
        S: state.S * scale,
        E: state.E * scale,
        I: state.I * scale,
        R: state.R * scale,
        D: state.D * scale,
      };
    }
  }

  return result;
}

/**
 * Verify population conservation: S + E + I + R + D = N at all times
 */
export function verifyPopulationConservation(
  trajectory: SEIRState[],
  expectedN: number,
  tolerance = 0.01
): { conserved: boolean; maxDrift: number; maxDriftDay: number } {
  let maxDrift = 0;
  let maxDriftDay = 0;

  for (let i = 0; i < trajectory.length; i++) {
    const total = trajectory[i].S + trajectory[i].E + trajectory[i].I + trajectory[i].R + trajectory[i].D;
    const drift = Math.abs(total - expectedN);
    if (drift > maxDrift) {
      maxDrift = drift;
      maxDriftDay = i;
    }
  }

  return {
    conserved: maxDrift < tolerance,
    maxDrift,
    maxDriftDay,
  };
}

/**
 * Compute R0 from parameters
 */
export function computeR0(beta: number, gamma: number, mu: number): number {
  return beta / (gamma + mu);
}

/**
 * Compute beta from desired R0
 */
export function computeBetaFromR0(r0: number, gamma: number, mu: number): number {
  return r0 * (gamma + mu);
}

File: predictive-epidemic-modeling-platform (1)/backend/app/core/seir_solver.py
"""
SEIR Differential Equations Solver

Extended SEIR model with mortality, vaccination, and climate modulation.
Uses SciPy's odeint for numerical integration.

Equations:
  dS/dt = -β*S*I/N - ν*S                  (Susceptible)
  dE/dt = β*S*I/N - σ*E                   (Exposed)
  dI/dt = σ*E - γ*I - μ*I                 (Infectious)
  dR/dt = γ*I + ν*S*(1-ε)                 (Recovered)
  dD/dt = μ*I                             (Deceased)

Where:
  β = transmission rate (climate-modulated)
  σ = incubation rate (1/incubation_period)
  γ = recovery rate (1/recovery_period)
  μ = mortality rate
  ν = vaccination rate
  ε = vaccine efficacy
"""

from typing import List, Dict, Tuple
import numpy as np
from scipy.integrate import odeint


def compute_beta_modulated(
    base_beta: float,
    avg_temp: float,
    humidity: float,
    mask_compliance: float,
    isolation_index: float,
) -> float:
    """Compute climate and intervention-modulated transmission rate."""
    # Temperature effect: peak transmission around 10-15°C
    temp_factor = 1 + 0.02 * np.exp(-((avg_temp - 12) ** 2) / 100)

    # Humidity effect: moderate humidity (40-60%) increases transmission
    humidity_factor = 1 + 0.1 * np.exp(-((humidity - 50) ** 2) / 2000)

    # Mask compliance reduces effective contact
    mask_factor = 1 - (mask_compliance / 100) * 0.6

    # Isolation reduces mixing
    isolation_factor = 1 - isolation_index * 0.7

    return max(0.01, base_beta * temp_factor * humidity_factor * mask_factor * isolation_factor)


def seir_deriv(
    y: np.ndarray,
    t: float,
    N: float,
    beta: float,
    sigma: float,
    gamma: float,
    mu: float,
    vaccine_rate: float,
    vaccine_efficacy: float,
    mobility_in: float = 0.0,
) -> np.ndarray:
    """
    Compute SEIRD derivatives.
    
    y = [S, E, I, R, D]
    """
    S, E, I, R, D = y
    v_rate = (vaccine_rate / 100) * 0.01  # Convert percentage to daily rate
    ve = vaccine_efficacy / 100

    # Force of infection
    contact_term = beta * S * I / N if N > 0 else 0
    vaccination_term = v_rate * S

    dS = -contact_term - vaccination_term + mobility_in
    dE = contact_term - sigma * E
    dI = sigma * E - gamma * I - mu * I
    dR = gamma * I + vaccination_term * (1 - ve)
    dD = mu * I

    return np.array([dS, dE, dI, dR, dD])


def solve_seir(
    population: float,
    initial_seir: Dict[str, float],
    params: Dict,
    days: int,
    avg_temp: float = 15.0,
    humidity: float = 60.0,
    mobility_flows: List[Tuple[int, float]] = None,
) -> List[Dict[str, float]]:
    """
    Solve SEIR model for a single node using odeint.
    
    Returns list of SEIR states for each day.
    """
    if mobility_flows is None:
        mobility_flows = []

    N = population
    y0 = np.array([
        initial_seir["S"],
        initial_seir["E"],
        initial_seir["I"],
        initial_seir["R"],
        initial_seir["D"],
    ])

    beta = compute_beta_modulated(
        params["beta"], avg_temp, humidity,
        params["mask_compliance"], params["isolation_index"],
    )

    # Time points (daily)
    t = np.linspace(0, days, days * 10 + 1)  # 10 substeps per day for accuracy

    # Build mobility interpolation
    mob_dict = {day: flow for day, flow in mobility_flows}

    def get_mobility(time_val: float) -> float:
        day = int(time_val)
        return mob_dict.get(day, 0.0) / 10  # Distribute over substeps

    # Solve with odeint
    def deriv_wrapper(y, time_val):
        mob = get_mobility(time_val)
        return seir_deriv(
            y, time_val, N, beta,
            params["sigma"], params["gamma"], params["mu"],
            params["vaccine_rate"], params["vaccine_efficacy"],
            mob,
        )

    solution = odeint(deriv_wrapper, y0, t, rtol=1e-8, atol=1e-10)

    # Extract daily snapshots
    daily_states = []
    step_size = len(solution) // (days + 1)
    if step_size == 0:
        step_size = 1

    for i in range(days + 1):
        idx = min(i * step_size, len(solution) - 1)
        state = solution[idx]
        # Clamp negatives
        state = np.maximum(state, 0)
        daily_states.append({
            "S": float(state[0]),
            "E": float(state[1]),
            "I": float(state[2]),
            "R": float(state[3]),
            "D": float(state[4]),
        })

    return daily_states


def verify_population_conservation(
    trajectory: List[Dict[str, float]],
    expected_n: float,
    tolerance: float = 0.01,
) -> Dict:
    """Verify S+E+I+R+D = N at all timepoints."""
    max_drift = 0
    max_drift_day = 0

    for i, state in enumerate(trajectory):
        total = state["S"] + state["E"] + state["I"] + state["R"] + state["D"]
        drift = abs(total - expected_n)
        if drift > max_drift:
            max_drift = drift
            max_drift_day = i

    return {
        "conserved": max_drift < tolerance,
        "max_drift": max_drift,
        "max_drift_day": max_drift_day,
    }


def compute_r0(beta: float, gamma: float, mu: float) -> float:
    """R0 = β / (γ + μ)"""
    return beta / (gamma + mu)


def compute_beta_from_r0(r0: float, gamma: float, mu: float) -> float:
    """β = R0 * (γ + μ)"""
    return r0 * (gamma + mu)
